util/log/context.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tobias Schottdorf
+
+package log
+
+import (
+	"fmt"
+	"strings"
+
+	"golang.org/x/net/context"
+)
+
+// Context wraps a context.Context into an object suitable for logging.
+type Context interface {
+	context.Context
+	With(kvs ...interface{}) Context
+
+	Infof(string, ...interface{})
+	Warningf(string, ...interface{})
+	Errorf(string, ...interface{})
+	Fatalf(string, ...interface{})
+}
+
+type logContext struct {
+	context.Context
+}
+
+// Background ...
+func Background() Context {
+	return Wrap(context.Background())
+}
+
+// Wrap ...
+func Wrap(ctx context.Context) Context {
+	return &logContext{ctx}
+}
+
+func (lc *logContext) With(kvs ...interface{}) Context {
+	ctx := lc.Context
+	l := len(kvs)
+	if l%2 != 0 {
+		panic("Add called with odd number of arguments")
+	}
+	for i := 1; i < l; i += 2 {
+		ctx = context.WithValue(ctx, kvs[i-1], kvs[i])
```

Comment:
We probably shouldn't mutate the struct; if we have values that are set at different times we should use separate structs. So this won't get us to a single struct but it organizes things a little bit (e.g. one struct for user/rpc context (method, client), one for addressing (node, store, range IDs), etc). Maybe that's more trouble than its worth and we're better off keeping the fields separate instead of figuring out how to group them.